tools/codegen/selective_build/operator.py

```diff
@@ -0,0 +1,117 @@
+from typing import Dict, List, Optional
+from dataclasses import dataclass
+
+# This class holds information about model metadata. Used to track
+# information about where the selective build operator list comes
+# from, and potentially which model each operator is used in.
+@dataclass
+class PyTorchModelMetadata:
+    name: str
+    version: str
+
+    @staticmethod
+    def from_yaml(data: Dict[str, object]) -> 'PyTorchModelMetadata':
+        name = data['name']
+        assert isinstance(name, str)
+
+        version = data['version']
+        assert isinstance(version, str)
+
+        return PyTorchModelMetadata(name, version)
+
+
+@dataclass(frozen=True)
+class SelectiveBuildOperator():
+    # The name of the operator. This includes the aten::, etc... prefix
+    # The operator name may or may not have the overload name. If this
+    # operator name does not specify an overload name, the way to determine
+    # if this entry refers to the family of operators with this base name
+    # or just the operator with this name is to look at the value of the
+    # 'include_all_overloads' flag in this class.
+    name: str
+
+    # True if this is a root operator (i.e. called directly from TorchScript, etc...)
+    is_root_operator: bool
+
+    # Is this operator used for on-device training? If True, then we need to
+    # use the information to generate code in VariableType_N.cpp for registration
+    # of training related operators
+    is_used_for_training: bool
+
+    # If True, it indicates that this operator instance (object) refers to an
+    # operator without the overload name and should apply to all overloads
+    # which have this operator name as the base name. This flag is applicable
+    # only for objects that have operator names without a DOT (period) character
+    # in them.
+    include_all_overloads: bool
+
+    # The list of models that use this operator.
+    models: Optional[List[PyTorchModelMetadata]]
```

Comment:
nit: probably don't need to be in this PR, but can we make this thing a bit generic? like debug info or something?
The reason is that people in the OSS community might not follow the same way to represent model metadata (name + version), so making it less specific is probably more flexible?

Another option is to move some logic to internal repo - but I think some functionality is useful for OSS eventually, e.g. it should be a common need to merge multiple selective build configs.

Reply:
@ljk53 Do you have a specific structure in mind for debug info? I think debug info is more generic as well. Maybe something simple like an info string would work right? Does that sound better? That way, the user retains the flexibility of formatting it in the way that they see fit. For example, for internal use-case, we can format the string as `model_name@version`, etc...

Reply:
Addressed in revision by replacing with an opaque string.